src/vs/workbench/contrib/chat/browser/chatViewTitle.ts

```diff
@@ -0,0 +1,184 @@
+
+import './chatViewTitle.css';
+
+import { h, append } from '../../../../base/browser/dom.js';
+import { Disposable } from '../../../../base/common/lifecycle.js';
+import { localize } from '../../../../nls.js';
+import { IConfigurationService } from '../../../../platform/configuration/common/configuration.js';
+import { IViewDescriptorService, IViewContainerModel } from '../../../common/views.js';
+import { ActivityBarPosition, LayoutSettings } from '../../../services/layout/browser/layoutService.js';
+import { IChatModel } from '../common/chatModel.js';
+
+export interface IChatViewTitleDelegate {
+	getCurrentModel(): IChatModel | undefined;
+	updatePrimaryTitle(title: string): void;
+	requestLayout(): void;
+}
+
+export class ChatViewTitleController extends Disposable {
+	private readonly viewContainerModel: IViewContainerModel | undefined;
+	private hasMultipleVisibleViews = false;
+	private showSecondaryTitleBar: boolean;
+	private secondarySideBarShowsLabels: boolean;
+	private currentPrimaryTitle: string;
+	private _secondaryTitleContainer: HTMLElement | undefined;
+	private _secondaryTitle: HTMLElement | undefined;
+
+	constructor(
+		private readonly viewId: string,
+		private readonly delegate: IChatViewTitleDelegate,
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IViewDescriptorService private readonly viewDescriptorService: IViewDescriptorService,
+	) {
+		super();
+		const viewContainer = this.viewDescriptorService.getViewContainerByViewId(this.viewId);
+		if (viewContainer) {
+			this.viewContainerModel = this.viewDescriptorService.getViewContainerModel(viewContainer);
+			this.hasMultipleVisibleViews = this.viewContainerModel.visibleViewDescriptors.length > 1;
+			this._register(this.viewContainerModel.onDidAddVisibleViewDescriptors(() => this.handleVisibleViewDescriptorsChanged()));
+			this._register(this.viewContainerModel.onDidRemoveVisibleViewDescriptors(() => this.handleVisibleViewDescriptorsChanged()));
+		}
+		this.showSecondaryTitleBar = this.shouldRenderSecondaryTitleBar();
+		this.secondarySideBarShowsLabels = this.computeSecondarySideBarLabelConfig();
+		this.currentPrimaryTitle = localize('chat', "Chat");
+		this._register(this.configurationService.onDidChangeConfiguration(e => {
+			if (e.affectsConfiguration(LayoutSettings.ACTIVITY_BAR_LOCATION)) {
+				this.handleActivityBarLocationChange();
+			}
+			if (e.affectsConfiguration('workbench.secondarySideBar.showLabels')) {
+				this.handleSecondarySideBarShowLabelsChange();
+			}
+		}));
+	}
+
+	render(parent: HTMLElement): void {
+		if (this._secondaryTitleContainer) {
+			return;
+		}
+		this._secondaryTitleContainer = h('div.chat-view-secondary-title-container').root;
+		this._secondaryTitle = h('span.chat-view-secondary-title-text').root;
+		append(this._secondaryTitleContainer, this._secondaryTitle);
+		append(parent, this._secondaryTitleContainer);
```

Comment:
@bpasero @roblourens I know this isn't quite the intended use of an `h` style function for building UI but felt just a touch wasteful to build the UI then go back and query select for the secondary title when I could just create/get a reference to the title at the same time.

If you'd prefer that I don't do this (and just build the whole UI and then query select) I can change this. 

_*As I think about it now, I suppose query selecting a single nested element is probably a wash in terms of performance anyway?_

Reply:
I pushed a commit to show how this is done since it's hard to explain

Reply:
That's right, and the real magic is that typescript knows about the name after the `@` so you get proper TS suggestions for the names. I couldn't even tell you how that works.

Reply:
Oooh very very cool. I'm definitely gonna check out how that works.